src/usb.rs

```diff
@@ -0,0 +1,69 @@
+//! USB peripheral.
+//!
+//! Provides the required implementation for use of the [`stm32-usbd`] crate.
+
+pub use stm32_usbd::UsbBus;
+
+use crate::gpio;
+use crate::gpio::gpioa::{PA11, PA12};
+use crate::rcc::{Enable, Reset};
+use crate::stm32::{RCC, USB};
+use core::fmt;
+use stm32_usbd::UsbPeripheral;
+
+/// Trait implemented by all pins that can be the "D-" pin for the USB peripheral
+pub trait DmPin: crate::Sealed {}
+
+/// Trait implemented by all pins that can be the "D+" pin for the USB peripheral
+pub trait DpPin: crate::Sealed {}
+
+impl DmPin for PA11<gpio::Alternate<{ gpio::AF14 }>> {}
+impl DpPin for PA12<gpio::Alternate<{ gpio::AF14 }>> {}
+
+pub struct Peripheral<Dm: DmPin, Dp: DpPin> {
+    /// USB register block
+    pub usb: USB,
+    /// Data negative pin
+    pub pin_dm: Dm,
+    /// Data positive pin
+    pub pin_dp: Dp,
+}
+
+impl<Dm, Dp> fmt::Debug for Peripheral<Dm, Dp>
+where
+    Dm: DmPin + fmt::Debug,
+    Dp: DpPin + fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Peripheral")
+            .field("usb", &"USB")
+            .field("pin_dm", &self.pin_dm)
+            .field("pin_dp", &self.pin_dp)
+            .finish()
+    }
+}
+
+// SAFETY: Implementation of Peripheral is thread-safe by using cricitcal sections to ensure
+// mutually exclusive access to the USB peripheral
+unsafe impl<Dm: DmPin, Dp: DpPin> Sync for Peripheral<Dm, Dp> {}
+
+// SAFETY: The peripheral has the same regiter blockout as the STM32 USBFS
+unsafe impl<Dm: DmPin + Send, Dp: DpPin + Send> UsbPeripheral for Peripheral<Dm, Dp> {
+    const REGISTERS: *const () = USB::ptr().cast::<()>();
+    const DP_PULL_UP_FEATURE: bool = true;
+    const EP_MEMORY: *const () = 0x4000_6000 as _;
+    const EP_MEMORY_SIZE: usize = 1024;
+    const EP_MEMORY_ACCESS_2X16: bool = true;
+
+    fn enable() {
+        critical_section::with(|_| unsafe {
```

Comment:
In other peripherals we pass a mutable reference to `Rcc` to ensure exclusive access. Would that work here?

Reply:
Oh wait, it is implementing a trait from outside this crate...

Reply:
What is the different between `critical_section::with` (requiring a new dependency) and `cortex_m::interrupt::free` as used in [src/serial/usart.rs](https://github.com/stm32-rs/stm32g4xx-hal/blob/main/src/serial/usart.rs#L194)?

Reply:
If there would be no downside, I would prefer to avoid the extra dependency